File: store/1.两数之和.c
/*
 * @lc app=leetcode.cn id=1 lang=c
 *
 * [1] 两数之和
 */

// @lc code=start

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int *twoSum(int *nums, int numsSize, int target, int *returnSize)
{
  int i, j, *arr, *res, len = 0, flag;

  arr = (int *)malloc(sizeof(int) * numsSize);
  res = (int *)malloc(sizeof(int) * 2);

  for (i = 0; i < numsSize; i++)
  {
    flag = 0;
    for (j = 0; j < len; j++)
    {
      if (nums[arr[j]] == (target - nums[i]))
      {
        flag = 1;
        break;
      }
    }
    if (flag == 1)
      break;
    else
      arr[len++] = i;
  }

  res[0] = i;
  res[1] = j;
  *returnSize = 2;
  return res;
}
// @lc code=end

File: store/1.两数之和.js
/*
 * @lc app=leetcode.cn id=1 lang=javascript
 *
 * [1] 两数之和
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function (nums, target) {
  // 写法一：
  // let map = new Map(),
  //   res = [];
  // nums.forEach((item, index) => {
  //   if (!map.has(item)) map.set(item, [index]);
  //   else map.get(item).push(index);
  // });
  // for (let item of nums) {
  //   const num = target - item;
  //   if (map.has(num)) {
  //     if (map.get(num)[0] !== map.get(item)[0]) {
  //       res = [map.get(item)[0], map.get(num)[0]];
  //       return res;
  //     }
  //     if (map.get(num)[0] === map.get(item)[0] && map.get(num).length > 1) {
  //       res = map.get(item);
  //       return res;
  //     }
  //   }
  // }
  // 写法二：
  // let map = new Map(),
  //   res = [];
  // nums.forEach((item, index) => map.set(item, index));
  // for (let [index, item] of nums.entries()) {
  //   const num = target - item;
  //   if (map.has(num)) {
  //     if (map.get(num) !== index) {
  //       res = [index, map.get(num)];
  //       return res;
  //     }
  //   }
  // }
  // 写法三：
  // let map = new Map(),
  //   res = [];
  // for (let [index, item] of nums.entries()) {
  //   const num = target - item;
  //   if (map.has(num)) {
  //     if (map.get(num) !== index) {
  //       res = [index, map.get(num)];
  //       return res;
  //     }
  //   } else map.set(item, index);
  // }

  const map = new Map();

  for (const [index, value] of nums.entries()) {
    const num = target - value;
    if (map.has(num)) {
      if (map.get(num) !== index) {
        return [index, map.get(num)];
      }
    } else {
      map.set(value, index);
    }
  }
};
// @lc code=end
